### ISSUE TITLE
Problem detecting network loss

### ISSUE THREAD

**Issue by RouquinBlanc** [aaugustin/websockets#112]
Hi Aymeric,

I'm having difficulties detecting a loss of connectivity on the websocket server. Basically, I would like my websocket to timeout quickly when we lose connectivity to the client.

my server side ws_handler is pretty simple, looking about like this:

``` python3
while True:
    data = # Get data from somewhere
    yield from websocket.send(data)
```

When the send buffer will be full, send() becomes a blocking call. If I disconnect the network cable from the client (reading on the websocket), I get quickly to a full send buffer, while TCP retries sending for ages.

I then modified the send to add a timeout myself:

``` python3
while True:
    data = # Get data from somewhere
    try:
        yield from asyncio.wait_for(websocket.send(data), TIMEOUT, loop=loop)
    except asyncio.TimeoutError:
         yield from websocket.close(reason='websocket_timeout')
```

But then the close remains stuck because it does not manage to send the OP_CLOSE frame (blocked on drain()).
I could also wrap the close() into a wait_for() too, but then the websocket would not be effectively closed. I have another coroutine waiting on the websocket.recv() part, and it would be good that it fails on a ConnectionClosed error.

The only way I manage to close the websocket without ending up blocked until TCP connection times out is by doing:

```
websocket.writer._transport.abort()
```

Which does not seem very great: I don't get a chance to set a code and a reason by doing so.

Could you advise me on the best way to achieve closing the connection, without waiting (or at least not too much), and if possible setting properly a code and reason?

Thanks
Jonathan

**Comment by aaugustin** [comment]
I believe this happens because disconnecting the network cable doesn't result in the TCP connection being closed.

Can you check if [connection_lost](https://github.com/aaugustin/websockets/blob/817a14fd3b85c0bdc4449bfe792098eaa2a45bfe/websockets/protocol.py#L647) gets called?

If yes, this looks like a bug.

If not, websockets has no way to know that the remote endpoint went away, so there isn't much I can do.

**Comment by RouquinBlanc** [comment]
connection_lost is not called, probably because the socket has not timed out yet (I guess?).

But:
- isn't there a timeout on the websocket itself, at higher level compared to TCP socket timeout?
- I can stick with wrapping the websocket.send(...) call into a wait_for(), but then when it times out, I want to close properly my websocket, with a proper reason, which currently I can't because the close() will get blocked itself. Do you provide a kind of websocket.force_close(reason='because I decided'), similar to aborting a transport?

**Comment by aaugustin** [comment]
There's a contradiction in your requirements: you want to "close properly [your] websocket" in a situation where you can't receive data from the other end. But closing properly a websocket requires receiving a close frame from the other end. So you can't "close properly" while the network cable is unplugged.

The documentation of [close()](http://websockets.readthedocs.io/en/stable/api.html#websockets.protocol.WebSocketCommonProtocol.close) suggests wrapping it in `ensure_future()` and forgetting about it. `close()` will complete in at most `websocket.timeout` seconds.

-- At this point I realized that `serve()` is missing a `timeout` keyword argument. I should fix that.

`websocket.timeout` isn't a timeout on the websocket connection. It's just a timeout on the closing handshake. I'm not sure what the logic for a more general timeout on the websocket would be. I can't see a reason why it should be different from the timeout of the underlying TCP connection.

Generally speaking I would recommend running a separate task or coroutine that calls `ping()` at regular interval, with a timeout, and terminates the connection if pings timeout. This will avoid mixing the data transfer logic and the connection supervision logic.

**Comment by RouquinBlanc** [comment]
You are right, I meant I mostly want to properly close the underlying socket, to release ressources.
```
The documentation of close() suggests wrapping it in ensure_future() and forgetting about it. close() will complete in at most websocket.timeout seconds.
```
As long as we agree that this is not currently the case, it's fine.
```python
    @asyncio.coroutine
    def close(self, code=1000, reason=''):
        if self.state == OPEN:
            # 7.1.2. Start the WebSocket Closing Handshake
            # 7.1.3. The WebSocket Closing Handshake is Started
            frame_data = serialize_close(code, reason)
            # !!!!!!!!!!! This will block until buffer is drained, or TCP socket times out...
            yield from self.write_frame(OP_CLOSE, frame_data)

        # If the connection doesn't terminate within the timeout, break out of
        # the worker loop.
        try:
            yield from asyncio.wait_for(
                self.worker, self.timeout, loop=self.loop)
        except asyncio.TimeoutError:
            self.worker.cancel()
```
Other thing: if it ends up cancelling the worker on close(), then any coroutine blocking on ```yield from websocket.recv()``` would get an undocumented CancelledError:
```python
        try:
            done, pending = yield from asyncio.wait(
                [next_message, self.worker],
                loop=self.loop, return_when=asyncio.FIRST_COMPLETED)
        except asyncio.CancelledError:
            # Handle the Task.cancel()
            next_message.cancel()
            raise  # !!!!! Shouldn't it raise a ConnectionClosed instead?
```

As for why having a timeout different of the TCP timeout, it's generally because the TCP timeout is too long, and on server side we want to quickly release resources when a client is gone, or any other reason specific reason.

**Comment by aaugustin** [comment]
Good catch! Definitely a bug. Thanks for narrowing it down.

**Comment by aaugustin** [comment]
You're also right about the wrong exception type.

**Comment by aaugustin** [comment]
Ugh I should really fix this.

**Comment by aaugustin** [comment]
Regarding the incorrect exception type, I don't think I can fix it, because `cancel()` is the only way to stop a running task and it doesn't allow injecting another exception type.

**Comment by aaugustin** [comment]
I've been working on a patch for unblocking `close()` when the output buffer is full and the connection stalled, but it causes a bunch of test failures I don't understand.

**Comment by RouquinBlanc** [comment]
For the exception type, can't you just `raise ConnectionClosed` instead of simply `raise` after cancelling the next_message task?

Otherwise if you think it it's not worth fixing, I still think it should be documented ;-)

**Comment by aaugustin** [comment]
Ah, right, I got confused with the exception injected into the worker coroutine by `cancel()`. I'll try again.

I also added a commit today that allows you to configure the buffer sizes. Lowering them may help with your issues of memory consumptions by connections that are timing out.

**Comment by aaugustin** [comment]
Reading this again, I think it may be fixed:

- I suspect https://github.com/python/asyncio/pull/280 was part of the problem; it was backported in WebSocketCommonProtocol some time ago;
- I rewrote connection termination entirely in #267; that work fixes #264 which I suspect is part of the problem as well; the result is much easier to reason about and I think it enforces timeouts correctly. The connection drops in at most 2 * timeout on the server side and 3 * timeout on the client side.

If you still care and you're still able to reproduce, trying #267 would be very useful. (Given how long this issue has been open, I wouldn't be surprised if you moved on to a library that doesn't have this issue.)

**Comment by RouquinBlanc** [comment]
Hey,

No I'm still using the initial trick of calling ```websocket.writer._transport.abort()``` to close the websocket. Harsh, but efficient ;-)

It's a bit crazy now at work to try and repro this issue, but I'm wondering if it would work. Need to be tested (with disconnecting cables during the test!)

**Comment by aaugustin** [comment]
Perhaps the easiest way to test this would be to wait for a release that includes #267, removing your workaround, and checking what happens. No hurry!

**Comment by aaugustin** [comment]
The problem still exists. I could reproduce it with the scripts below.

On macOS, apply the comments at the beginning of the server script to make sure the operating system doesn't allow buffers so large that it's painful to fill them. There's probably something similar on Linux.

Even with this settings and after reducing ``websockets`` buffers, I still need to ship about 1MB of data to fill the OS buffers. I'm not sure why.

To reproduce the issue, start the server script in a terminal and then the client script in another terminal.

The server hits `!!! TIMEOUT 1 !!! ` after 1 second and blocks until the client terminates 19 seconds later.

**server112.py**

```python
#!/usr/bin/env python

# Reduce TCP buffers (original value: 131072):
# sudo sysctl -w net.inet.tcp.recvspace=4096
# sudo sysctl -w net.inet.tcp.sendspace=4096

# Disable TCP buffer autotuning (original value: 1):
# sudo sysctl -w net.inet.tcp.doautosndbuf=0
# sudo sysctl -w net.inet.tcp.doautorcvbuf=0

import asyncio
import logging
import websockets

logging.basicConfig(
    format='%(asctime)s  %(levelname)-8s %(message)s',
    level=logging.DEBUG,
)

logger = logging.getLogger()

async def test112(websocket, path):
    try:
        for i in range(2 ** 12):            # 4096 messages
            data = '{:0256b}'.format(i)     # 256 bytes each, total = 1MB
            await asyncio.wait_for(websocket.send(data), timeout=1)
    except asyncio.TimeoutError:
        logger.info("!!! TIMEOUT 1 !!!")
        try:
            await websocket.close()
        except asyncio.TimeoutError:
            logger.info("!!! TIMEOUT 2 !!!")

start_server = websockets.serve(
    test112, 'localhost', 8765,
    timeout=2, read_limit=1024, write_limit=1024)

asyncio.get_event_loop().run_until_complete(start_server)
asyncio.get_event_loop().run_forever()
```

***client112.py***

```python
#!/usr/bin/env python

import asyncio
import logging
import websockets

logging.basicConfig(
    format='%(asctime)s  %(levelname)-8s %(message)s',
    level=logging.DEBUG,
)

async def test112():
    async with websockets.connect(
        'ws://localhost:8765',
        read_limit=1024, write_limit=1024, max_queue=4
    ) as websocket:
        await asyncio.sleep(20)
        async for _ in websocket:
            pass

asyncio.get_event_loop().run_until_complete(test112())
```
